plugins/osdn/netid/controller/controller.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/openshift/openshift-sdn/plugins/osdn/netid/vnidallocator"
+	"github.com/openshift/origin/pkg/client/cache"
+	"github.com/openshift/origin/pkg/controller"
+)
+
+type VnidController struct {
+	vnid   vnidallocator.Interface
+	client kclient.NamespaceInterface
+	Queue  *cache.EventQueue
+
+	// globalNamespaces are the namespaces that have access to all pods in the cluster and vice versa.
+	globalNamespaces sets.String
+}
+
+func NewVnidController(vnida vnidallocator.Interface, client kclient.NamespaceInterface, globalNamespaces []string) *VnidController {
+	return &VnidController{
+		vnid:             vnida,
+		client:           client,
+		globalNamespaces: sets.NewString(globalNamespaces...),
+	}
+}
+
+// Create creates a vnid controller
+func (c *VnidController) Create() controller.RunnableController {
+	if c.Queue == nil {
+		lw := &kcache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return c.client.List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return c.client.Watch(options)
+			},
+		}
+		q := cache.NewEventQueue(kcache.MetaNamespaceKeyFunc)
+		kcache.NewReflector(lw, &kapi.Namespace{}, q, 10*time.Minute).Run()
+		c.Queue = q
+	}
+	return c
+}
+
+// Run begins processing resources from Queue asynchronously.
+func (c *VnidController) Run() {
+	go utilwait.Forever(func() { c.handleOne(c.Queue.Pop()) }, 0)
+}
+
+// handleOne processes resource with Handle.
+func (c *VnidController) handleOne(eventType watch.EventType, resource interface{}, err error) {
+	if err != nil {
+		glog.Errorf("Failed to pop namespace from event queue: %v", err)
+		return
+	}
+
+	ns := resource.(*kapi.Namespace)
+	switch eventType {
+	case watch.Added, watch.Modified:
+		err = c.addOrUpdate(ns)
+	case watch.Deleted:
+		err = c.delete(ns)
+	}
+	if err != nil {
+		glog.Errorf("Failed to apply vnid allocation, event-type: %v, namespace: %v, error: %v", eventType, ns, err)
+	}
+}
+
+// retryCount is the number of times to retry on a conflict when updating a namespace
+const retryCount = 2
+
+// addOrUpdate allocates network id(VNID) for the namespace if needed
+func (c *VnidController) addOrUpdate(ns *kapi.Namespace) error {
+	tx := &tx{}
+	defer tx.Rollback()
+
+	var id uint
+	userRequestedVnid := false
+	value, err := netid.GetRequestedVNID(ns)
+	if err == nil {
+		id = value
+		userRequestedVnid = true
+		netid.DeleteRequestedVNID(ns)
+	} else if err != netid.ErrorVNIDNotFound {
+		return err
+	} else if _, err = netid.GetVNID(ns); err == nil {
+		// Nothing to do in this case
+		// VNID already allocated and namespace does not want new VNID
+		return nil
+	}
+
+	name := ns.ObjectMeta.Name
+	if userRequestedVnid {
+		// Only accept global VNID or VNIDs that are already allocated
+		if (id != netid.GlobalVNID) && !c.vnid.Has(id) {
+			return fmt.Errorf("requested netid %d not allocated", id)
+		}
+	} else {
+		// Do vnid allocation
+		if c.globalNamespaces.Has(name) {
```

Comment:
Why are you taking this as a config value when it could be stored on the namespace?